src/vae.py

```diff
@@ -0,0 +1,238 @@
+        self.project = self._linear(self.latent_size, self.q_in, relu=False)
+        self.decoder = nn.Sequential(
+            self._deconv(self.n_kernels, self.n_kernels // 2),
+            self._deconv(self.n_kernels // 2, self.n_kernels // 4),
+            self._deconv(self.n_kernels // 4, self.n_channels),
+            nn.Sigmoid(),
+        )
+
+    def create_encoder(self, n_channels, latent_size):
+        self.encoder = nn.Sequential(
+            self._conv(self.n_channels, self.n_kernels // 4),
+            self._conv(self.n_kernels // 4, self.n_kernels // 2),
+            self._conv(self.n_kernels // 2, self.n_kernels),
+        )
+
+        seq_dicts = [
+            {
+                "C_in": self.n_channels,
+                "C_out": self.n_kernels // 4,
+                "N": 1,
+                "kernel_size": self.kernel_size,
+                "stride": self.stride,
+                "padding": self.padding,
+            },
+            {
+                "C_in": self.n_kernels // 4,
+                "C_out": self.n_kernels // 2,
+                "N": 1,
+                "kernel_size": self.kernel_size,
+                "stride": self.stride,
+                "padding": self.padding,
+            },
+            {
+                "C_in": self.n_kernels // 2,
+                "C_out": self.n_kernels,
+                "N": 1,
+                "kernel_size": self.kernel_size,
+                "stride": self.stride,
+                "padding": self.padding,
+            },
+        ]
+        self.output_shapes = GetConv2dSequentialShape(
+            input_shape=(1, self.n_channels) + self.img_shape, li=seq_dicts
+        )
+        self.q_in = int(np.prod(self.output_shapes[-1][1:]))
+        # A layer to arrive at the mus and the sigmas
+
+        self.q_mu = nn.Linear(self.q_in, latent_size)
+        self.q_vars = nn.Linear(self.q_in, latent_size)
+
+    def forward(self, x) -> Tuple[Tuple[T, T], T]:
+        # First, derive a latent representation
+        x_hat = self.encoder(x)
+        x_hat.shape
+        # Then, from the latent variables' PDF of Q(z), sample the means and the vars
+        mus, log_sigmas = self.Q(x_hat)
+        # Once the mus and the
+        z_sampled = self.z(mus, log_sigmas)
+        z_sampled.shape
+        z_projected = self.project(z_sampled).view(
+            -1, self.n_kernels, self.output_shapes[-1][2], self.output_shapes[-1][3]
+        )  # Project and reshape back to an image
+
+        output = self.decoder(z_projected)
+
+        return (mus, log_sigmas), output
+
+    def Q(self, x_hat):
+        # Q is the latent variables' PDF
+
+        x_hat = x_hat.view(-1, self.q_in)
+        mu, log_sigma = self.q_mu(x_hat), self.q_vars(x_hat)
+        return mu, log_sigma
+
+    def z(self, mu, log_sigma):
+        # Sample from z
+        sigma = log_sigma.mul(0.5).exp_()
+        # Here we sample from a normal distribution
+        #
+        ret = torch.randn(mu.shape)
```

Comment:
How are the gradients able to propagate through this step?

Reply:
It is not supposed to, but rather through `sigma` and `mu` - it is the [reparametrization trick](https://stats.stackexchange.com/questions/199605/how-does-the-reparameterization-trick-for-vaes-work-and-why-is-it-important)

Reply:
Ah yeah they flow through the `mul` and `add_` below :) thanks!